### PR TITLE
:hammer: (db) migrate Dataset and Source to knex

### PR DESCRIPTION
**Pull request by sophiamersmann**
- Makes the `Dataset` and `Source` TypeORM classes obsolete by switching db access to knex
- This PR does not do any clean-up, e.g. SQL queries in the router files are not moved, although they ultimately should live in the `db` folder

- I added two helper functions that can be used as drop-in replacements for `queryMysql` and `mysqlFirst` that are often used in our codebase, `knexRaw` and `knexRawFirst` (both live in `db.ts`)

- Routes that I could easily test are tested, but not all of the code is!

@danyx23 let me know if this is how you imagined it, or if I should do anything differently.

@coderabbitai ignore

### CODE REVIEW

**Comment by sophiamersmann** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3131?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `master`
  * **PR #3054** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3054?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #3056** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3056?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #3064** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3064?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #3131** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3131?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
        * **PR #3074** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3074?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/owid/owid-grapher/3131?utm_source=stack-comment).

File: adminSiteServer/apiRouter.ts [hunk 1]
```diff
@@ -1949,27 +1954,24 @@ apiRouter.put("/datasets/:datasetId", async (req: Request, res: Response) => {
...
-            commitName: res.locals.user.fullName,
-            commitEmail: res.locals.user.email,
-        })
-    } catch (err) {
-        logErrorAndMaybeSendToBugsnag(err, req)
-        // Continue
-    }
+        try {
+            await syncDatasetToGitRepo(t, datasetId, {
+                oldDatasetName: dataset.name!,
```
Comment (sophiamersmann) [inline adminSiteServer/apiRouter.ts:1966]: The TypeORM code just assumed that a dataset `name` was given. That's why I felt it was safe to sometimes ignore a possible undefined value via Typescript's `!`. Let me know if you prefer for this to be handled properly.

File: adminSiteServer/apiRouter.ts [hunk 2]
```diff
@@ -2450,16 +2451,16 @@ apiRouter.post("/posts/:postId/unlinkGdoc", async (req: Request) => {
 
 apiRouter.get("/sources/:sourceId.json", async (req: Request) => {
     const sourceId = expectInt(req.params.sourceId)
-    const source = await db.mysqlFirst(
+    const source = await db.knexRawFirst<Record<string, any>>(
```
Comment (sophiamersmann) [inline adminSiteServer/apiRouter.ts:2454]: When it wasn't particularly important, I didn't use specific types. Let me know if you prefer to have strict types everywhere.

**Review by danyx23 — CHANGES_REQUESTED** [review]
Thanks a lot for test driving the knex approach! This looks very good overall. There are two things that I would change on the db-types PR learning from this and two things I'd ask you to change.

The two changs I'd like to see to this PR is:
- not make the knex instance optional in the knexRaw and knexRawFirst helper functions
- and to wrap all API functions we touch in a transaction scope even if we only read data as we go through them. The reasoning here is that we want reads to be consistent and it's probably better to just implement a standard pattern everywhere rather than having to evaluate if it is necessary each time.

The two things I'll change on the other PR are:
- bump the knex version so that we can create readyOnly transactions which are a bit more performant than generic ones
- add a migration to make name on datasets mandatory rather than nullable

**Review by danyx23 — CHANGES_REQUESTED** [review]
Thanks a lot for test driving the knex approach! This looks very good overall. There are two things that I would change on the db-types PR learning from this and two things I'd ask you to change.

The two changs I'd like to see to this PR is:
- not make the knex instance optional in the knexRaw and knexRawFirst helper functions
- and to wrap all API functions we touch in a transaction scope even if we only read data as we go through them. The reasoning here is that we want reads to be consistent and it's probably better to just implement a standard pattern everywhere rather than having to evaluate if it is necessary each time.

The two things I'll change on the other PR are:
- bump the knex version so that we can create readyOnly transactions which are a bit more performant than generic ones
- add a migration to make name on datasets mandatory rather than nullable

**Review by sophiamersmann — COMMENTED** [review]
Good to go I think!

Before the `name` field was made non-nullable in the db, I added a check to make sure a dataset `name` is given. We could get rid of that now :)

**Pull request merged**